programs/perpetuals/src/instructions/add_locked_stake.rs

```diff
@@ -0,0 +1,369 @@
+                },
+            )?;
+        }
+
+        (
+            stake_amount_with_reward_multiplier,
+            stake_amount_with_lm_reward_multiplier,
+        )
+    };
+
+    // transfer newly staked tokens to Stake PDA
+    msg!("Transfer tokens");
+    {
+        perpetuals.transfer_tokens_from_user(
+            ctx.accounts.funding_account.to_account_info(),
+            ctx.accounts.staking_token_account.to_account_info(),
+            ctx.accounts.owner.to_account_info(),
+            ctx.accounts.token_program.to_account_info(),
+            params.amount,
+        )?;
+    }
+
+    // Give governing power to the Stake owner
+    {
+        // Apply voting multiplier related to locking period
+        let voting_power = math::checked_as_u64(math::checked_div(
+            math::checked_mul(params.amount, staking_option.vote_multiplier as u64)? as u128,
+            Perpetuals::BPS_POWER,
+        )?)?;
+
+        perpetuals.add_governing_power(
+            ctx.accounts.transfer_authority.to_account_info(),
+            ctx.accounts.owner.to_account_info(),
+            ctx.accounts.owner.to_account_info(),
+            ctx.accounts
+                .governance_governing_token_owner_record
+                .to_account_info(),
+            ctx.accounts.governance_token_mint.to_account_info(),
+            ctx.accounts.governance_realm.to_account_info(),
+            ctx.accounts.governance_realm_config.to_account_info(),
+            ctx.accounts
+                .governance_governing_token_holding
+                .to_account_info(),
+            ctx.accounts.governance_program.to_account_info(),
+            voting_power,
+            None,
+            true,
+        )?;
+    }
+
+    // update Cortex data
+    {
+        cortex.next_staking_round.total_stake = math::checked_add(
+            cortex.next_staking_round.total_stake,
+            stake_amount_with_reward_multiplier,
+        )?;
+
+        cortex.next_staking_round.lm_total_stake = math::checked_add(
+            cortex.next_staking_round.lm_total_stake,
+            stake_amount_with_lm_reward_multiplier,
+        )?;
+    }
+
+    // If auto claim thread is paused, resume it
+    {
+        if ctx.accounts.stakes_claim_cron_thread.paused {
+            clockwork_sdk::cpi::thread_resume(CpiContext::new_with_signer(
+                ctx.accounts.clockwork_program.to_account_info(),
+                clockwork_sdk::cpi::ThreadResume {
+                    authority: ctx.accounts.staking_thread_authority.to_account_info(),
+                    thread: ctx.accounts.stakes_claim_cron_thread.to_account_info(),
+                },
+                &[&[
+                    STAKING_THREAD_AUTHORITY_SEED,
+                    ctx.accounts.owner.key().as_ref(),
+                    &[ctx.accounts.staking.thread_authority_bump],
+                ]],
+            ))?;
+        }
+    }
+
+    Ok(())
+}
```

Comment:
It's complexe but it reads well and nothing seems out of place so far, neat work